### PR TITLE
ConsoleView: Recursively get steps for all stages (and child stages)

### LINKED ISSUE #22: PipelineConsoleView: children of a node as shown as sequential steps, and not lower level tree items.
The tree returned by the `tree` REST API call should be a tree representation, but it is a graph (it's just a copy of the `graph` rest call).
FlowGraphTable does something similar when ordering the rows for its table:
https://github.com/jenkinsci/workflow-support-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java#L126

### PR DESCRIPTION
**Pull request by timbrown5**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

I realised the current code only loops through the top level of stages when trying to find steps. This fixes the code so  it gets the steps for each stage (but recursive down to each child).

Fixes #22 
> The FrontEnd doesn't seem to call steps API for the lowest level stages (they are always empty).

### CODE REVIEW

**Comment by timbrown5** [comment]
The code in this PR is currently broken. I have uplaoded it in the hope that someone can see what I am doing wrong.

I am trying to record the stage we are interested in in enclosingNode:
https://github.com/timbrown5/pipeline-graph-view-plugin/blob/addStepsToAllLevels/src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java#L149

What I thiink is happening is that this is too aggressive and get set to a child of the parallel block when we want it to he the parallel block itself (meaning the key in the map is wrong).

Here is where we add the step to the stageStepMap:
https://github.com/timbrown5/pipeline-graph-view-plugin/blob/addStepsToAllLevels/src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java#L188

**Comment by timja** [comment]
Is there a good example pipeline to reproduce with?

**Comment by timbrown5** [comment]
> Is there a good example pipeline to reproduce with?
The screenshot above is using the pipeline from this comment (under 'Nested').
https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/15#issuecomment-865661205

**Comment by timja** [comment]
This simple pipeline is now logged out of order:
https://github.com/timja-org/junit-attachments-test/blob/master/Jenkinsfile

**Comment by timja** [comment]
Looks like the list of steps is just in reverse (ideally needs some test coverage)

**Comment by timbrown5** [comment]
> Looks like the list of steps is just in reverse (ideally needs some test coverage), not sure if we need to reverse, adjust parsing, or sort by node id?

That's a good point. I think if I push the steps to a stack and then empty the stack when I get to a start node it will fix that issue (and hopefully the other issue).

**Comment by timbrown5** [comment]
- Updated the visitor to use a stack and which fixed the ordering. 
- Updated the API so it only creates one instance of the StepVisitor - it was to aid debugging, but should actually make it more efficient too.
- Update StepVisitor to store a list of pending blocks (if we hit and end, we add to pending, when we hit start we pop from pending). This allowed me to only add steps to a stage if it's parent wasn't a parallel stage - which looked like it's resolved the issue.

Screenshot 2021-07-25 at 10 44 23
 
Here is the log from both of the visitors:
<details>
  <summary>Full log</summary>
2021-07-25 09:43:13.302+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Non-Parallel Stage' with id '6' is type 'STAGE'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Parallel Stage' with id '11' is type 'STAGE'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Branch A' with id '15' is type 'PARALLEL'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Branch B' with id '16' is type 'PARALLEL'.
2021-07-25 09:43:13.303+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Branch C' with id '17' is type 'PARALLEL'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Skipped stage' with id '59' is type 'STAGE'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Nested 1' with id '25' is type 'STAGE'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Nested 2' with id '41' is type 'STAGE'.
2021-07-25 09:43:13.304+0000 [id=106]   INFO    i.j.p.p.utils.PipelineGraphApi#lambda$createTree$10: Stage 'Nested 3' with id '47' is type 'STAGE'.
2021-07-25 09:43:13.502+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 59 - true - false - false .
2021-07-25 09:43:13.503+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 59 - true - false - false .
2021-07-25 09:43:13.503+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.503+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 0.
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 11 - true - false - false .
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchEnd: Node ID: 17 - false - true - false .
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 23 - true - false - false .
2021-07-25 09:43:13.504+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 47 - true - false - false .
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 49 to stack.
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 48 to stack.
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 47 - true - false - false .
2021-07-25 09:43:13.506+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '48' to '47'.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '49' to '47'.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.507+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 41 - true - false - false .
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 43 to stack.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 42 to stack.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 41 - true - false - false .
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '42' to '41'.
2021-07-25 09:43:13.508+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '43' to '41'.
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 25 - true - false - false .
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 37 to stack.
2021-07-25 09:43:13.509+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 36 to stack.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 25 - true - false - false .
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '36' to '25'.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '37' to '25'.
2021-07-25 09:43:13.510+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 23 - true - false - false .
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Not pushing steps to stage in handleChunkDone as parent is parallel block.
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Node ID: 17 - false - true - false .
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Pushing steps to stage in parallelBranchStart.
2021-07-25 09:43:13.511+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 0.
2021-07-25 09:43:13.512+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchEnd: Node ID: 16 - false - true - false .
2021-07-25 09:43:13.516+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 21 - true - false - false .
2021-07-25 09:43:13.516+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 30 to stack.
2021-07-25 09:43:13.516+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 29 to stack.
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 21 - true - false - false .
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Not pushing steps to stage in handleChunkDone as parent is parallel block.
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Node ID: 16 - false - true - false .
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Pushing steps to stage in parallelBranchStart.
2021-07-25 09:43:13.517+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '29' to '16'.
2021-07-25 09:43:13.518+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '30' to '16'.
2021-07-25 09:43:13.518+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.518+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchEnd: Node ID: 15 - false - true - false .
2021-07-25 09:43:13.527+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 19 - true - false - false .
2021-07-25 09:43:13.527+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 27 to stack.
2021-07-25 09:43:13.528+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 26 to stack.
2021-07-25 09:43:13.529+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 19 - true - false - false .
2021-07-25 09:43:13.529+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Not pushing steps to stage in handleChunkDone as parent is parallel block.
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Node ID: 15 - false - true - false .
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#parallelBranchStart: Pushing steps to stage in parallelBranchStart.
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '26' to '15'.
2021-07-25 09:43:13.530+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '27' to '15'.
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 2.
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 11 - true - false - false .
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.531+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 0.
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkEnd: Node ID: 6 - true - false - false .
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#atomNode: Pushing step: 7 to stack.
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#chunkStart: Node ID: 6 - true - false - false .
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#handleChunkDone: Pushing steps to stage in handleChunkDone.
2021-07-25 09:43:13.532+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Adding step '7' to '6'.
2021-07-25 09:43:13.533+0000 [id=114]   INFO    i.j.p.p.u.PipelineStepVisitor#pushStageStepsToMap: Clearing 1.
</details>

File: src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java [hunk 1]
```diff
@@ -56,12 +62,13 @@ public void getSteps(StaplerRequest req, StaplerResponse rsp) throws IOException
         String nodeId = req.getParameter("nodeId");
         if (nodeId != null) {
             LOGGER.log(Level.FINE, "getSteps was passed nodeId '" + nodeId + "'.");
-            PipelineStepApi stepApi = new PipelineStepApi(target, nodeId);
             ObjectMapper mapper = new ObjectMapper();
-            LOGGER.log(Level.FINE, "Steps: '" + mapper.writeValueAsString(stepApi.getSteps()) + "'.");
-            rsp.getWriter().append(mapper.writeValueAsString(stepApi.getSteps()));
+            PipelineStepList steps = stepApi.getSteps(nodeId);
+            LOGGER.log(Level.FINE, "Steps: '" + mapper.writeValueAsString(steps) + "'.");
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:53]: should be wrapped in either a supplier or an `if LOGGER.isLoggable` pattern so the object isn't converted to a string when not needed

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApi.java [hunk 2]
```diff
@@ -22,41 +22,45 @@
...
     private static final Logger LOGGER = Logger.getLogger(PipelineStepApi.class.getName());
     private transient WorkflowRun run;
-    private transient FlowNode node;
     private final String nodeId;
 
+    private static PipelineStepVisitor builder;
+    private static Object mutex = new Object();
 
-    public PipelineStepApi(WorkflowRun run, String nodeId) throws java.io.IOException {
+    public PipelineStepApi(WorkflowRun run, String nodeId) {
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApi.java:19]: node id is assigned but never accessed is there a reason for that?

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineGraphApi.java [hunk 3]
```diff
@@ -120,11 +127,85 @@ public PipelineGraph createGraph() {
...
+        List<PipelineStageInternal> stages = getPipelineNodes();
+        List<Integer> topLevelStageIds = new ArrayList<>();
+
+        // id => stage
+        Map<Integer, PipelineStageInternal> stageMap = stages.stream()
+                .collect(Collectors.toMap(PipelineStageInternal::getId, stage -> stage, (u, v) -> u, LinkedHashMap::new));
+
+        Map<Integer, List<Integer>> stageToChildrenMap = new HashMap<>();
+
+        List<Integer> stageIds = new ArrayList<>();
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineGraphApi.java:144]: this list doesn't appear to be used? (other than adding to it)

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java [hunk 4]
```diff
@@ -56,14 +56,20 @@
...
 
     private final Map<String,FlowNodeWrapper> stepMap = new HashMap<>();
+    private final Map<String,List<FlowNodeWrapper>> stageStepMap = new HashMap<>();
+
+    private final ArrayDeque<FlowNode> pendingBlocks = new ArrayDeque<>();
 
     private boolean stageStepsCollectionCompleted = false;
 
     private boolean inStageScope;
+    private boolean inParallelBranchScope;
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java:64]: unused?

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java [hunk 5]
```diff
@@ -56,14 +56,20 @@
...
+    private boolean parentnIsParallelBranch;
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepVisitor.java:65]: unused and typo on name?

**Review by timja — CHANGES_REQUESTED** [review]
I've done some code cleanup, can you take a look at the rest of my comments? 

Is there anything else outstanding on this. I've re-tested and the issues I mentioned before are fixed

File: src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApi.java (hunk 2 above)
Comment (timbrown5) [reply src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApi.java:19]: An artefact of the was I was trying to get the steps before.

**Comment by timbrown5** [comment]
Sorry managed to make my changes when in a detached head, so pushed the one instance of the branch.

**Comment by timbrown5** [comment]
> Is there anything else outstanding on this. I've re-tested and the issues I mentioned before are fixed

I would like to add a test for the StepAPI before this is landed. Your cleanup inadvertently broke the parsing (I did the same thing last night when trying to clean up locally). Adding a test should stop this from happening again - I started this on a different branch before so should be able to revive it.

**Comment by timja** [comment]
> > Is there anything else outstanding on this. I've re-tested and the issues I mentioned before are fixed
> 
> I would like to add a test for the StepAPI before this is landed. Your cleanup inadvertently broke the parsing (I did the same thing last night when trying to clean up locally). Adding a test should stop this from happening again - I started this on a different branch before so should be able to revive it.

interesting, didn't think I changed any logic, just places where my IDE warned the condition could not be true

**Comment by timbrown5** [comment]
> interesting, didn't think I changed any logic, just places where my IDE warned the condition could not be true
Spotbugs was warning about this
`pendingBlocks.pop()` which got removed (but is needed to track which stage we should be adding the steps to.

**Comment by timbrown5** [comment]
> I would like to add a test for the StepAPI before this is landed. Your cleanup inadvertently broke the parsing (I did the same thing last night when trying to clean up locally). Adding a test should stop this from happening again - I started this on a different branch before so should be able to revive it.

I have added tests to check that the PipelineGraphAPI.createTree and PipelineStepsAPI.getSteps work as expected in our simple use case.

File: src/test/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApiTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        assertThat(steps.get(1).getName(), is("Set stage result to unstable"));
+
+        steps = api.getSteps(failureID).getSteps();
+        assertThat(steps, hasSize(2));
+        assertThat(steps.get(0).getName(), is("Set stage result to unstable"));
+        assertThat(steps.get(1).getName(), is("Error signal"));
+    }
+
+    @Test
+    public void complexParallelParallelBranchesHaveCorrectSteps() throws Exception {
```
Comment (timja) [inline src/test/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineStepApiTest.java:57]:
```suggestion
    public void complexParallelBranchesHaveCorrectSteps() throws Exception {
```

**Review by timja — APPROVED** [review]

**Pull request merged**